Refactors Preference Classes

shobhitagarwal1612 opened the pull request:
Addresses #441 

- Adds PreferenceFragments for each PreferenceActivity to reduce deprecated methods
- Adds ToastUtils in utilities 
- Deletes AppPreferenceActivity

shobhitagarwal1612 committed:
Moves AboutPreferencesFragment out of AboutPreferencesActivity

shobhitagarwal1612 committed:
Uses PreferenceFragment in AdminPreferencesActivity

shobhitagarwal1612 committed:
Creates AggregatePreferenceFragment from AggregatePreferenceActivity

shobhitagarwal1612 committed:
Adds GooglePreferencesFragment
Adds PreferencesFragment
Adds OtherPreferencesFragment

shobhitagarwal1612 committed:
Deletes AppPreferenceActivity

shobhitagarwal1612 committed:
Replaces pa with pf in DisabledPreferencesRemover

shobhitagarwal1612 committed:
Adds ToastUtils in utilities
Removes unused imports

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.java:
```diff
@@ -0,0 +1,19 @@
+
+import org.odk.collect.android.application.Collect;
+
+/**
```
I can’t speak for the project, but I find these noisy and wish Studio wouldn’t put them in by default. Git tells us what we need to know.

dcbriccetti replied:
It’s not visible here, but I’m referring to the “Created by x” excretions.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.java:
```diff
@@ -0,0 +1,19 @@
+
+public class ToastUtils {
+    public static void shortDuration(String message){
+        Toast.makeText(Collect.getInstance(),message, Toast.LENGTH_SHORT).show();
```
Space after comma. A better name might be makeShortText. Perhaps there should a third, private method that factors out the common code.

shobhitagarwal1612 replied:
Agreed

dcbriccetti replied:
Correction: showShortText might be better. I missed the show() at the end. What do you think?

shobhitagarwal1612 replied:
What if we use showShortToast?

dcbriccetti replied:
OK with me. It’s sort of implied by the “ToastUtils.” but it would add clarity for someone who statically imported the methods.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/preferences/AggregatePreferencesFragment.java:
```diff
@@ -176,9 +176,7 @@ public boolean onPreferenceChange(Preference preference, Object newValue) {
-                    Toast.makeText(getActivity().getApplicationContext(),
-                            R.string.url_error, Toast.LENGTH_SHORT)
-                            .show();
+                    ToastUtils.shortDuration(String.valueOf(R.string.url_error));
```
An overloaded ToastUtils.makeShort/LongText method that takes the “R” value would save a lot of “String.valueOf”s in the callers.

dcbriccetti commented on collect_app/src/main/res/xml/other_preferences.xml:
```diff
@@ -20,5 +19,4 @@
             android:title="@string/submission_url" />
     </PreferenceCategory>
-   
 </PreferenceScreen>
```
I would add the newline here at the end of the file if for no other reason than to remove the warning.

shobhitagarwal1612 committed:
Removed "Created by" comment in added files
Renamed methods in ToastUtils
Added a private method in ToastUtils to factor out common code

shobhitagarwal1612 committed:
Adds overloaded function to accept "string resources" directly without using extra wrapper (String.ValueOf())
Adds two final variables to ToastUtils

shobhitagarwal1612 committed:
Replaced displaying Toasts with ToastUtils class in rest of the classes

shobhitagarwal1612 committed:
Adds catching NotFoundException in ToastUtils

shobhitagarwal1612 committed:
removes redundant call to getString()

lognaturel commented:
@dcbriccetti @shobhitagarwal1612 Let me know when this is ready for my review, please!

shobhitagarwal1612 commented:
@lognaturel Please review the changes

nribeka commented on collect_app/src/main/java/org/odk/collect/android/preferences/AdminPreferencesFragment.java:
```diff
@@ -0,0 +1,123 @@
+    public boolean onPreferenceClick(Preference preference) {
+
+        if (preference.getKey().equals(KEY_CHANGE_ADMIN_PASSWORD)) {
+            AlertDialog.Builder b = new AlertDialog.Builder(getActivity());
```
I'm not a fan of one char variable. But it could be just a preference for some.

dcbriccetti replied:
I agree with @nribeka on this. Especially when it’s in scope for so long. But, did this change create that name? Most of this change is moving things around.

lognaturel replied:
Agreed.

lognaturel replied:
Yes, good point by @dcbriccetti. We should be picky style-wise about new code but it's ok to leave old things alone to minimize invasive changes.

shobhitagarwal1612 replied:
@dcbriccetti The name of this builder variable was not changed by me. But I will change that to `builder`

dcbriccetti replied:
No, I think a small consensus formed that we wouldn’t change it.

nribeka commented on collect_app/src/main/java/org/odk/collect/android/preferences/PreferencesActivity.java:
```diff
@@ -14,111 +14,56 @@
     public static final String INTENT_KEY_ADMIN_MODE = "adminMode";
     protected static final int IMAGE_CHOOSER = 0;
-
+    private static final String t = "PreferencesActivity";
```
Use TAG and you can use the simple class name instead of hard coding the String.

`PreferencesActivity.class.getSimpleName()`

lognaturel replied:
We can't do that because of proguard.

lognaturel replied:
https://medium.com/@elye.project/the-danger-of-using-class-getsimplename-as-tag-for-fragment-5cdf3a35bfe2#.yiz5an5eb

But I do agree that the constant name should be TAG. The codebase is inconsistent so we can just change this when we do a style check across all classes.

nribeka replied:
Ah the dang proguard. I guess I need to change my code then.

shobhitagarwal1612 replied:
@lognaturel @nribeka So we can safely use `PreferenceActivity.class.getSimpleName()` for the `PreferenceActivity` TAG?

lognaturel replied:
Unfortunately, no.

nribeka replied:
Yeah, proguard will obfuscate this class name and then the log will ended up with "A", or "B", or "C" instead of "PreferenceActivity".

nribeka commented on collect_app/src/main/java/org/odk/collect/android/preferences/PreferencesFragment.java:
```diff
@@ -0,0 +1,402 @@
+
+public class PreferencesFragment extends PreferenceFragment implements Preference.OnPreferenceChangeListener {
+    public static final String INTENT_KEY_ADMIN_MODE = "adminMode";
+    private static final String t = "PreferencesActivity";
```
I would do the same here with TAG and simple name.

shobhitagarwal1612 replied:
:+1:

lognaturel replied:
Ah, I see. Sorry, I didn't respond to this one. Doesn't work here either.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.java:
```diff
@@ -0,0 +1,42 @@
+
+public class ToastUtils {
+
+    private static final int LENGTH_SHORT = Toast.LENGTH_SHORT;
```
No need for these. “void toaster(...)” can use the Toast ones.

shobhitagarwal1612 replied:
:+1:

shobhitagarwal1612 replied:
Should I use `Toast` instead of `toaster`?

lognaturel replied:
Please remove these constants. You should just refer directly to `Toast.LENGTH_xx`. There is no value to saving their values locally.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/utilities/ToastUtils.java:
```diff
@@ -0,0 +1,42 @@
+        Toast.makeText(Collect.getInstance(), message, duration).show();
+    }
+
+    private static void toaster(int messageResource, int duration) {
```
Method names should be a verb or verb phrase.

dcbriccetti replied:
showToast would be a good name.

shobhitagarwal1612 committed:
Renames variables to more logical names
Renames toaster to Toast in ToastUtils

shobhitagarwal1612 committed:
reverts back value of TAG

shobhitagarwal1612 committed:
Renames Toast to showToast

shobhitagarwal1612 committed:
Removes try/catch block for the unchecked exception in ToastUtils

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/preferences/PreferencesFragment.java:
```diff
@@ -0,0 +1,402 @@
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.odk.collect.android.preferences.PreferenceKeys.ARRAY_INDEX_GOOGLE_MAPS;
```
In the case of static constant imports like these, I think it's better to do a single wildcard import.

lognaturel reviewed (changes requested):
In the future, please do try to keep your pull requests as focused as possible. In this case, you should only have been touching preference-related classes and in fact it might have been better to do separate PRs for admin and general settings. When a PR includes unrelated changes across many files, errors are more likely to sneak in!

shobhitagarwal1612 committed:
Removes constants from ToastUtils
Optimzed imports

shobhitagarwal1612 commented:
@lognaturel I am sorry for the inconvenience caused. Will strictly follow to keep PRs `focused` from now on. I have done all the requested changes.
